modules/sitemaps/sitemap-builder.php

```diff
@@ -0,0 +1,1458 @@
+require_once dirname( __FILE__ ) . '/sitemap-finder.php';
+require_once dirname( __FILE__ ) . '/sitemap-state.php';
+
+if ( defined( 'WP_DEBUG' ) && ( true === WP_DEBUG ) ) {
+	require_once dirname( __FILE__ ) . '/sitemap-logger.php';
+}
+
+/**
+ * The Jetpack_Sitemap_Builder object handles the construction of
+ * all sitemap files (except the XSL files, which are handled by
+ * Jetpack_Sitemap_Stylist.) Other than the constructor, there are
+ * only two public functions: build_all_sitemaps and news_sitemap_xml.
+ *
+ * @since 4.7.0
+ */
+class Jetpack_Sitemap_Builder {
+
+	/**
+	 * Librarian object for storing and retrieving sitemap data.
+	 *
+	 * @access private
+	 * @since 4.7.0
+	 * @var $librarian Jetpack_Sitemap_Librarian
+	 */
+	private $librarian;
+
+	/**
+	 * Logger object for reporting debug messages.
+	 *
+	 * @access private
+	 * @since 4.7.0
+	 * @var $logger Jetpack_Sitemap_Logger
+	 */
+	private $logger;
+
+	/**
+	 * Finder object for dealing with sitemap URIs.
+	 *
+	 * @access private
+	 * @since 4.7.0
+	 * @var $finder Jetpack_Sitemap_Finder
+	 */
+	private $finder;
+
+	/**
+	 * Construct a new Jetpack_Sitemap_Builder object.
+	 *
+	 * @access public
+	 * @since 4.7.0
+	 */
+	public function __construct() {
+		$this->librarian = new Jetpack_Sitemap_Librarian();
+		$this->finder = new Jetpack_Sitemap_Finder();
+
+		if ( defined( 'WP_DEBUG' ) && ( true === WP_DEBUG ) ) {
+			$this->logger = new Jetpack_Sitemap_Logger();
+		}
+
+		update_option(
+			'jetpack_sitemap_post_types',
+			/**
+			 * The array of post types to be included in the sitemap.
+			 *
+			 * Add your custom post type name to the array to have posts of
+			 * that type included in the sitemap. The default array includes
+			 * 'page' and 'post'.
+			 *
+			 * The result of this filter is cached in an option, 'jetpack_sitemap_post_types',
+			 * so this filter only has to be applied once per generation.
+			 *
+			 * @since 4.7.0
+			 */
+			apply_filters(
+				'jetpack_sitemap_post_types',
+				array( 'post', 'page' )
+			)
+		);
+
+		return;
+	}
+
+	/**
+	 * Update the sitemap.
+	 *
+	 * @since 4.7.0
+	 */
+	public function update_sitemap() {
+		if ( defined( 'WP_DEBUG' ) && ( true === WP_DEBUG ) ) {
+			$this->logger->report( '-- Updating...' );
+		}
+
+		for ( $i = 1; $i <= JP_SITEMAP_UPDATE_SIZE; $i++ ) {
+			$this->build_next_sitemap_file();
+		}
+
+		if ( defined( 'WP_DEBUG' ) && ( true === WP_DEBUG ) ) {
+			$this->logger->report( '-- ...done for now.' );
+			$this->logger->time();
+		}
+	}
+
+	/**
+	 * Generate the next sitemap file.
+	 *
+	 * Reads the most recent state of the sitemap generation phase and
+	 * constructs the next file.
+	 *
+	 * @since 4.7.0
+	 */
+	private function build_next_sitemap_file() {
+		// Get the most recent state.
+		$state = Jetpack_Sitemap_State::check_out();
+
+		// Do nothing if the state is locked.
+		if ( false === $state ) {
+			return;
+		}
+
+		// Otherwise, branch on the sitemap-type key of $state.
```

Comment:
Unless we need the `===`, a `switch` statement would be clearer here. (`switch` does comparisons with `==`.)